File: rxjava-core/src/main/java/rx/Notification.java
/**
 * Copyright 2014 Netflix, Inc.
 * 
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 * 
 * http://www.apache.org/licenses/LICENSE-2.0
 * 
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package rx;

/**
 * An object representing a notification sent to an {@link Observable}.
 * 
 * For the Microsoft Rx equivalent see: http://msdn.microsoft.com/en-us/library/hh229462(v=vs.103).aspx
 */
public class Notification<T> {

    private final Kind kind;
    private final Throwable throwable;
    private final T value;

    private static final Notification<Void> ON_COMPLETED = new Notification<Void>(Kind.OnCompleted, null, null);

    public static <T> Notification<T> createOnNext(T t) {
        return new Notification<T>(Kind.OnNext, t, null);
    }

    public static <T> Notification<T> createOnError(Throwable e) {
        return new Notification<T>(Kind.OnError, null, e);
    }

    @SuppressWarnings("unchecked")
    public static <T> Notification<T> createOnCompleted() {
        return (Notification<T>) ON_COMPLETED;
    }

    @SuppressWarnings("unchecked")
    public static <T> Notification<T> createOnCompleted(Class<T> type) {
        return (Notification<T>) ON_COMPLETED;
    }

    private Notification(Kind kind, T value, Throwable e) {
        this.value = value;
        this.throwable = e;
        this.kind = kind;
    }

    /**
     * Retrieves the exception associated with an onError notification.
     * 
     * @return Throwable associated with an onError notification.
     */
    public Throwable getThrowable() {
        return throwable;
    }

    /**
     * Retrieves the data associated with an onNext notification.
     * 
     * @return The data associated with an onNext notification.
     */
    public T getValue() {
        return value;
    }

    /**
     * Retrieves a value indicating whether this notification has a value.
     * 
     * @return a value indicating whether this notification has a value.
     */
    public boolean hasValue() {
        return isOnNext() && value != null;
    }

    /**
     * Retrieves a value indicating whether this notification has an exception.
     * 
     * @return a value indicating whether this notification has an exception.
     */
    public boolean hasThrowable() {
        return isOnError() && throwable != null;
    }

    /**
     * Retrieves the kind of the notification: OnNext, OnError, OnCompleted
     * 
     * @return the kind of the notification: OnNext, OnError, OnCompleted
     */
    public Kind getKind() {
        return kind;
    }

    public boolean isOnError() {
        return getKind() == Kind.OnError;
    }

    public boolean isOnCompleted() {
        return getKind() == Kind.OnCompleted;
    }

    public boolean isOnNext() {
        return getKind() == Kind.OnNext;
    }

    public void accept(Observer<? super T> observer) {
        if (isOnNext()) {
            observer.onNext(getValue());
        } else if (isOnCompleted()) {
            observer.onCompleted();
        } else if (isOnError()) {
            observer.onError(getThrowable());
        }
    }

    public static enum Kind {
        OnNext, OnError, OnCompleted
    }

    @Override
    public String toString() {
        StringBuilder str = new StringBuilder("[").append(super.toString()).append(" ").append(getKind());
        if (hasValue())
            str.append(" ").append(getValue());
        if (hasThrowable())
            str.append(" ").append(getThrowable().getMessage());
        str.append("]");
        return str.toString();
    }

    @Override
    public int hashCode() {
        int hash = getKind().hashCode();
        if (hasValue())
            hash = hash * 31 + getValue().hashCode();
        if (hasThrowable())
            hash = hash * 31 + getThrowable().hashCode();
        return hash;
    }

    @Override
    public boolean equals(Object obj) {
        if (obj == null)
            return false;
        if (this == obj)
            return true;
        if (obj.getClass() != getClass())
            return false;
        Notification<?> notification = (Notification<?>) obj;
        if (notification.getKind() != getKind())
            return false;
        if (hasValue() && !getValue().equals(notification.getValue()))
            return false;
        if (hasThrowable() && !getThrowable().equals(notification.getThrowable()))
            return false;
        return true;
    }
}

File: language-adaptors/rxjava-scala/src/main/scala/rx/lang/scala/Notification.scala
/**
 * Copyright 2013 Netflix, Inc.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package rx.lang.scala

/**
 * Emitted by Observables returned by [[rx.lang.scala.Observable.materialize]].
 */
sealed trait Notification[+T] {
  private [scala] val asJavaNotification: rx.Notification[_ <: T]

  override def equals(that: Any): Boolean = that match {
    case other: Notification[_] => asJavaNotification.equals(other.asJavaNotification)
    case _ => false
  }
  override def hashCode(): Int = asJavaNotification.hashCode()

  /**
   * Invokes the function corresponding to the notification.
   *
   * @param onNext
   *               The function to invoke for an [[rx.lang.scala.Notification.OnNext]] notification.
   * @param onError
   *               The function to invoke for an [[rx.lang.scala.Notification.OnError]] notification.
   * @param onCompleted
   *               The function to invoke for an [[rx.lang.scala.Notification.OnCompleted]] notification.
   */
  def accept[R](onNext: T=>R, onError: Throwable=>R, onCompleted: ()=>R): R = {
    this match {
      case Notification.OnNext(value)  => onNext(value)
      case Notification.OnError(error) => onError(error)
      case Notification.OnCompleted  => onCompleted()
    }
  }

  def apply[R](onNext: T=>R, onError: Throwable=>R, onCompleted: ()=>R): R =
     accept(onNext, onError, onCompleted)

  /**
   * Invokes the observer corresponding to the notification
   *
   * @param observer
   *                 The observer that to observe the notification
   */
  def accept(observer: Observer[T]): Unit = {
    this match {
      case Notification.OnNext(value)  => observer.onNext(value)
      case Notification.OnError(error) => observer.onError(error)
      case Notification.OnCompleted  => observer.onCompleted()
    }
  }

  def apply(observer: Observer[T]): Unit = accept(observer)
}

/**
 * Provides pattern matching support and constructors for Notifications.
 * 
 * Example:
 * {{{
 * import Notification._
 * Observable(1, 2, 3).materialize.subscribe(n => n match {
 *   case OnNext(v)     => println("Got value " + v)
 *   case OnCompleted() => println("Completed")
 *   case OnError(err)  => println("Error: " + err.getMessage)
 * })
 * }}}
 */
object Notification {

  private [scala] def apply[T](n: rx.Notification[_ <: T]): Notification[T] = n.getKind match {
    case rx.Notification.Kind.OnNext => new OnNext(n)
    case rx.Notification.Kind.OnCompleted => OnCompleted
    case rx.Notification.Kind.OnError => new OnError(n)
  }
  
  // OnNext, OnError, OnCompleted are not case classes because we don't want pattern matching
  // to extract the rx.Notification

  object OnNext {

    /**
     * Constructor for onNext notifications.
     *
     * @param value
     * The item passed to the onNext method.
     */
    def apply[T](value: T): Notification[T] = {
      Notification(rx.Notification.createOnNext[T](value))
    }

    /**
     * Extractor for onNext notifications.
     * @param notification
     *                     The [[rx.lang.scala.Notification]] to be destructed.
     * @return
     *         The item contained in this notification.
     */
    def unapply[U](notification: Notification[U]): Option[U] = notification match {
      case onNext: OnNext[U] => Some(onNext.value)
      case _ => None
    }
  }

  class OnNext[+T] private[scala] (val asJavaNotification: rx.Notification[_ <: T]) extends Notification[T] {
    def value: T = asJavaNotification.getValue
    override def toString = s"OnNext($value)"
  }

  object OnError {

    /**
     * Constructor for onError notifications.
     *
     * @param error
     * The exception passed to the onNext method.
     */
    def apply[T](error: Throwable): Notification[T] = {
      Notification(rx.Notification.createOnError[T](error))
    }

    /**
     * Destructor for onError notifications.
     *
     * @param notification
     *                     The [[rx.lang.scala.Notification]] to be deconstructed
     * @return
     *         The `java.lang.Throwable` value contained in this notification.
     */
    def unapply[U](notification: Notification[U]): Option[Throwable] = notification match {
      case onError: OnError[U] => Some(onError.error)
      case _ => None
    }
  }

  class OnError[+T] private[scala] (val asJavaNotification: rx.Notification[_ <: T]) extends Notification[T] {
    def error: Throwable = asJavaNotification.getThrowable
    override def toString = s"OnError($error)"
  }

  object OnCompleted extends Notification[Nothing] {
    override def toString = "OnCompleted"
    val asJavaNotification = rx.Notification.createOnCompleted[Nothing]()
  }

}

